src/main/ruby/truffleruby/core/io.rb

```diff
@@ -1242,19 +1244,40 @@ def read_to_separator_with_limit
         break unless available > 0
 
         if count = @buffer.find(@separator)
-          bytes = Primitive.min(count, wanted)
-          str << @buffer.shift(bytes)
+          bytes_to_read = Primitive.min(count, wanted)
+          s = @buffer.shift(bytes_to_read)
 
-          str = IO.read_encode(@io, str)
+          # We need to be careful matching against multi-byte separators since the
+          # `str` value is being built up progressively.
+          #
+          # If the separator is only a single byte wide and we found it in the buffer
+          # then `count` must be the correct position because there's no way for a single
+          # byte read to be split up; it doesn't matter what `str` contains.
+          #
+          # If the separator is multiple bytes, then we look at `str`. If it's empty,
+          # then the buffer trivially must contain the separator. If not, however, we
+          # must scan the entire `str` after appending the buffer to see if the separator
+          # pattern appears earlier than the position detected by `count`.
+          if str.empty? || separator_byte_size == 1
+            unless str.empty?
+              s.prepend(str)
+              str.clear
+            end
 
-          str.chomp!(@separator) if @chomp
-          $. = @io.__send__(:increment_lineno)
-          @buffer.discard @skip if @skip
+            s = IO.read_encode(@io, s)
 
-          yield str
+            s.chomp!(@separator) if @chomp
+            $. = @io.__send__(:increment_lineno)
+            @buffer.discard @skip if @skip
 
-          str = +''
-          wanted = limit
+            yield s
+
+            wanted = limit
+
+            next
+          else
+            str << s
+          end
```

Comment:
After `str << s` we should probably adjust `wanted` as well, e.g. `wanted -= bytes_to_read` as far as `wanted = @limit - str.bytesize` is supposed to always be true.

What if `str` is longer than `@limit` now?